crates/sql_provider_datafusion/src/dbconnection.rs

```diff
@@ -10,14 +10,60 @@ pub mod postgresconn;
 pub type Error = Box<dyn std::error::Error + Send + Sync>;
 type Result<T, E = Error> = std::result::Result<T, E>;
 
-pub trait DbConnection<T, P>: Send {
+pub trait SyncDbConnection<T, P>: DbConnection<T, P> {
     fn new(conn: T) -> Self
     where
         Self: Sized;
     fn get_schema(&self, table_reference: &TableReference) -> Result<SchemaRef>;
     fn query_arrow(&self, sql: &str, params: &[P]) -> Result<SendableRecordBatchStream>;
     fn execute(&self, sql: &str, params: &[P]) -> Result<u64>;
+}
+
+#[async_trait::async_trait]
+pub trait AsyncDbConnection<T, P>: DbConnection<T, P> + Sync {
+    fn new(conn: T) -> Self
+    where
+        Self: Sized;
+    async fn get_schema(&self, table_reference: &TableReference) -> Result<SchemaRef>;
+    async fn query_arrow(&self, sql: &str, params: &[P]) -> Result<SendableRecordBatchStream>;
+    async fn execute(&self, sql: &str, params: &[P]) -> Result<u64>;
+}
 
+pub trait DbConnection<T, P>: Send {
     fn as_any(&self) -> &dyn Any;
     fn as_any_mut(&mut self) -> &mut dyn Any;
+
+    fn as_sync(&self) -> Option<&dyn SyncDbConnection<T, P>> {
+        None
+    }
+    fn as_async(&self) -> Option<&dyn AsyncDbConnection<T, P>> {
+        None
+    }
+}
+
+pub async fn get_schema<T, P>(
```

Comment:
putting this into DbConnnection result in a cycle dependency between DbConnection and the Async/Sync.

It's not the best, but still hiding the details from the caller.